File: ios/Classes/MetawearBoardChannel.swift
import Flutter
import MetaWear
import MetaWearCpp

public class MetawearBoardChannel {

  public let device: MetaWear
  private var channel: FlutterMethodChannel?
  private var connectionStateHandler: EventChannelHandler?

  private var registrar: FlutterPluginRegistrar?

  private var accelerationSignal: OpaquePointer? = nil
  private var magneticFieldSignal: OpaquePointer? = nil
  private var gyroscopeSignal: OpaquePointer? = nil
  private var quaternionSignal: OpaquePointer? = nil

  public var id: String {
    return device.mac ?? device.peripheral.identifier.uuidString
  }

  init(_ device: MetaWear, _ registrar: FlutterPluginRegistrar) {
    self.device = device
    self.channel = nil
    self.connectionStateHandler = nil

    DispatchQueue.main.async {
      self.channel = FlutterMethodChannel(
        name: "\(NAMESPACE)/metawear/\(device.peripheral.identifier.uuidString)",
        binaryMessenger: registrar.messenger()
      )
      self.connectionStateHandler = EventChannelHandler(
        id: "\(NAMESPACE)/metawear/\(device.peripheral.identifier.uuidString)/state",
        messenger: registrar.messenger()
      )
      self.channel?.setMethodCallHandler(self.handle)
    }
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    print("print handle(): \(call.method)")
    switch call.method {
    case "connect":
      self.connect(result: result)
    case "disconnect":
      self.disconnect(result: result)
    case "isConnected":
      result(device.isConnectedAndSetup)
    case "battery":
      self.getBattery(result: result)
    case "deviceInfo":
      self.getDeviceInfo(result: result)
    case "model":
      device.readModelNumber().continueWith { t in
        if let error = t.error {
          print("Error reading model number: \(error.localizedDescription)")
          result(nil)
        } else {
          print("Model number: \(t.result!)")
          result(t.result!)
        }
      }
    case "startCorrectedAcceleration":
      self.startCorrectedAcceleration(arguments: call.arguments, result: result)
    case "startCorrectedAngularVelocity":
      self.startCorrectedAngularVelocity(arguments: call.arguments, result: result)
    case "startCorrectedMagneticField":
      self.startCorrectedMagneticField(arguments: call.arguments, result: result)
    case "startQuaternion":
      self.startQuaternion(arguments: call.arguments, result: result)
    case "stop":
      self.stop(result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  public func connect(result: @escaping FlutterResult) {
    print("Connect()")
    device.connectAndSetup().continueWith { t in
      t.result?.continueWith { task in
        print("Lost connection")
        try self.connectionStateHandler?.success(event: [
          "connected": false, "reason": "Lost connection",
        ])
      }

      if let error = t.error {
        print("Error: \(error.localizedDescription)")
        try self.connectionStateHandler?.success(event: [
          "connected": false, "reason": error.localizedDescription,
        ])
        result(nil)
      } else {
        print("Connected")
        try self.connectionStateHandler?.success(event: ["connected": true])
        result(true)
      }

    }
  }

  public func disconnect(result: @escaping FlutterResult) {
    print("Disconnect()")
    device.cancelConnection()
    result(true)

  }

  public func getBattery(result: @escaping FlutterResult) {
    // TODO: Implement battery
    result(FlutterMethodNotImplemented)
  }

  public func getDeviceInfo(result: @escaping FlutterResult) {
    let info: DeviceInformation? = self.device.info

    result([
      "manufacturer": info?.manufacturer,
      "modelNumber": info?.modelNumber,
      "serialNumber": info?.serialNumber,
      "firmwareRevision": info?.firmwareRevision,
      "hardwareRevision": info?.hardwareRevision,
    ])
  }

  private func configureSensor(arguments: [String: String]) {

    let mode = arguments["mode"]
    let accRange = arguments["accRange"]
    let gyroRange = arguments["gyroRange"]

    mbl_mw_sensor_fusion_set_mode(
      device.board,
      mode == "IMU_PLUS"
        ? MBL_MW_SENSOR_FUSION_MODE_IMU_PLUS
        : mode == "COMPASS"
          ? MBL_MW_SENSOR_FUSION_MODE_COMPASS
          : mode == "M4G"
            ? MBL_MW_SENSOR_FUSION_MODE_M4G
            : mode == "NDOF"
              ? MBL_MW_SENSOR_FUSION_MODE_NDOF
              : mode == "SLEEP" ? MBL_MW_SENSOR_FUSION_MODE_SLEEP : MBL_MW_SENSOR_FUSION_MODE_NDOF)
    mbl_mw_sensor_fusion_set_acc_range(
      device.board,
      accRange == "AR_16G"
        ? MBL_MW_SENSOR_FUSION_ACC_RANGE_16G
        : accRange == "AR_8G"
          ? MBL_MW_SENSOR_FUSION_ACC_RANGE_8G
          : accRange == "AR_4G"
            ? MBL_MW_SENSOR_FUSION_ACC_RANGE_4G
            : accRange == "AR_2G"
              ? MBL_MW_SENSOR_FUSION_ACC_RANGE_2G : MBL_MW_SENSOR_FUSION_ACC_RANGE_16G)
    mbl_mw_sensor_fusion_set_gyro_range(
      device.board,
      gyroRange == "GR_2000DPS"
        ? MBL_MW_SENSOR_FUSION_GYRO_RANGE_2000DPS
        : gyroRange == "GR_1000DPS"
          ? MBL_MW_SENSOR_FUSION_GYRO_RANGE_1000DPS
          : gyroRange == "GR_500DPS"
            ? MBL_MW_SENSOR_FUSION_GYRO_RANGE_500DPS
            : gyroRange == "GR_250DPS"
              ? MBL_MW_SENSOR_FUSION_GYRO_RANGE_250DPS : MBL_MW_SENSOR_FUSION_GYRO_RANGE_2000DPS)
    mbl_mw_sensor_fusion_write_config(self.device.board)
  }

  public func startCorrectedAcceleration(arguments: Any?, result: @escaping FlutterResult) {
    self.accelerationSignal = mbl_mw_sensor_fusion_get_data_signal(
      self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_ACC)

    mbl_mw_datasignal_subscribe(self.accelerationSignal, bridge(obj: self)) {
      (context, obj) in
      let data: MblMwCorrectedCartesianFloat = obj!.pointee.valueAs()
      let source: MetawearBoardChannel = bridge(ptr: context!)
      DispatchQueue.main.async {
        print(Double(data.x), Double(data.y), Double(data.z))
        source.channel?.invokeMethod(
          "onCorrectedAcceleration",
          arguments: [
            "x": Double(data.x),
            "y": Double(data.y),
            "z": Double(data.z),
          ])
      }
    }

    self.configureSensor(arguments: arguments as! [String: String])

    mbl_mw_sensor_fusion_enable_data(self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_ACC)
    mbl_mw_sensor_fusion_start(self.device.board)

    result(true)
  }

  public func startCorrectedAngularVelocity(arguments: Any?, result: @escaping FlutterResult) {
    self.gyroscopeSignal = mbl_mw_sensor_fusion_get_data_signal(
      self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_GYRO)

    mbl_mw_datasignal_subscribe(self.gyroscopeSignal, bridge(obj: self)) {
      (context, obj) in
      let data: MblMwCorrectedCartesianFloat = obj!.pointee.valueAs()
      let source: MetawearBoardChannel = bridge(ptr: context!)
      DispatchQueue.main.async {
        print(Double(data.x), Double(data.y), Double(data.z))
        source.channel?.invokeMethod(
          "onCorrectedAngularVelocity",
          arguments: [
            "x": Double(data.x),
            "y": Double(data.y),
            "z": Double(data.z),
          ])
      }
    }

    self.configureSensor(arguments: arguments as! [String: String])

    mbl_mw_sensor_fusion_enable_data(self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_GYRO)
    mbl_mw_sensor_fusion_start(self.device.board)

    result(true)
  }

  public func startCorrectedMagneticField(arguments: Any?, result: @escaping FlutterResult) {
    self.magneticFieldSignal = mbl_mw_sensor_fusion_get_data_signal(
      self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_MAG)

    mbl_mw_datasignal_subscribe(self.magneticFieldSignal, bridge(obj: self)) {
      (context, obj) in
      let data: MblMwCorrectedCartesianFloat = obj!.pointee.valueAs()
      let source: MetawearBoardChannel = bridge(ptr: context!)
      DispatchQueue.main.async {
        print(Double(data.x), Double(data.y), Double(data.z))
        source.channel?.invokeMethod(
          "onCorrectedMagneticField",
          arguments: [
            "x": Double(data.x),
            "y": Double(data.y),
            "z": Double(data.z),
          ])
      }
    }

    self.configureSensor(arguments: arguments as! [String: String])

    mbl_mw_sensor_fusion_enable_data(self.device.board, MBL_MW_SENSOR_FUSION_DATA_CORRECTED_MAG)
    mbl_mw_sensor_fusion_start(self.device.board)

    result(true)
  }

  public func startQuaternion(arguments: Any?, result: @escaping FlutterResult) {
    self.quaternionSignal = mbl_mw_sensor_fusion_get_data_signal(
      self.device.board, MBL_MW_SENSOR_FUSION_DATA_QUATERNION)

    mbl_mw_datasignal_subscribe(self.quaternionSignal, bridge(obj: self)) {
      (context, obj) in
      let data: MblMwQuaternion = obj!.pointee.valueAs()
      let source: MetawearBoardChannel = bridge(ptr: context!)
      DispatchQueue.main.async {
        print(Double(data.x), Double(data.y), Double(data.z), Double(data.w))
        source.channel?.invokeMethod(
          "onQuaternion",
          arguments: [
            "w": Double(data.w),
            "x": Double(data.x),
            "y": Double(data.y),
            "z": Double(data.z),
          ])
      }
    }

    mbl_mw_sensor_fusion_enable_data(self.device.board, MBL_MW_SENSOR_FUSION_DATA_QUATERNION)
    mbl_mw_sensor_fusion_start(self.device.board)

    result(true)
  }

  public func stop(result: @escaping FlutterResult) {
    print("Stop()")
    mbl_mw_sensor_fusion_stop(self.device.board)
    mbl_mw_sensor_fusion_clear_enabled_mask(self.device.board)
    if self.accelerationSignal != nil {
      mbl_mw_datasignal_unsubscribe(self.accelerationSignal!)
    }
    if self.gyroscopeSignal != nil {
      mbl_mw_datasignal_unsubscribe(self.gyroscopeSignal!)
    }
    if self.magneticFieldSignal != nil {
      mbl_mw_datasignal_unsubscribe(self.magneticFieldSignal!)
    }
    if self.quaternionSignal != nil {
      mbl_mw_datasignal_unsubscribe(self.quaternionSignal!)
    }
    result(true)
  }

}

File: android/src/main/java/ai/healables/metawear_dart/MetawearBoardChannel.java
package ai.healables.metawear_dart;

import android.app.Activity;
import android.content.Context;

import com.mbientlab.metawear.DeviceInformation;
import com.mbientlab.metawear.MetaWearBoard;
import com.mbientlab.metawear.Observer;
import com.mbientlab.metawear.Route;
import com.mbientlab.metawear.module.Accelerometer;
import com.mbientlab.metawear.data.Quaternion;
import com.mbientlab.metawear.module.Settings;
import com.mbientlab.metawear.module.SensorFusionBosch;

import java.util.HashMap;

import bolts.Continuation;
import bolts.Task;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;
import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import android.util.Log;

public class MetawearBoardChannel implements MethodChannel.MethodCallHandler {
    private final BinaryMessenger messenger;
    private final MetaWearBoard board;
    private final Activity activity;

    private final MethodChannel deviceChannel;
    private final StreamHandler stateHandler;

    private SensorFusionBosch sensor;

    private SensorFusionBosch getSensor(MethodCall methodCall) {
        if (this.sensor == null) {
            this.sensor = board.getModule(SensorFusionBosch.class);

            SensorFusionBosch.Mode mode = methodCall.argument("mode") == "IMU_PLUS" ? SensorFusionBosch.Mode.IMU_PLUS
                    : methodCall.argument("mode") == "COMPASS" ? SensorFusionBosch.Mode.COMPASS
                            : methodCall.argument("mode") == "M4G" ? SensorFusionBosch.Mode.M4G
                                    : methodCall.argument("mode") == "NDOF" ? SensorFusionBosch.Mode.NDOF
                                            : methodCall.argument("mode") == "SLEEP" ? SensorFusionBosch.Mode.SLEEP
                                                    : SensorFusionBosch.Mode.NDOF;
            SensorFusionBosch.AccRange accRange = methodCall.argument("accRange") == "AR_16G"
                    ? SensorFusionBosch.AccRange.AR_16G
                    : methodCall.argument("accRange") == "AR_8G" ? SensorFusionBosch.AccRange.AR_8G
                            : methodCall.argument("accRange") == "AR_4G" ? SensorFusionBosch.AccRange.AR_4G
                                    : methodCall.argument("accRange") == "AR_2G" ? SensorFusionBosch.AccRange.AR_2G
                                            : SensorFusionBosch.AccRange.AR_16G;
            SensorFusionBosch.GyroRange gyroRange = methodCall.argument("gyroRange") == "GR_2000DPS"
                    ? SensorFusionBosch.GyroRange.GR_2000DPS
                    : methodCall.argument("gyroRange") == "GR_1000DPS" ? SensorFusionBosch.GyroRange.GR_1000DPS
                            : methodCall.argument("gyroRange") == "GR_500DPS" ? SensorFusionBosch.GyroRange.GR_500DPS
                                    : methodCall.argument("gyroRange") == "GR_250DPS"
                                            ? SensorFusionBosch.GyroRange.GR_250DPS
                                            : SensorFusionBosch.GyroRange.GR_2000DPS;

            this.sensor.configure()
                    .mode(mode)
                    .accRange(accRange)
                    .gyroRange(gyroRange)
                    .commit();
        }
        return this.sensor;
    }

    public MetawearBoardChannel(
            final BinaryMessenger messenger,
            final MetaWearBoard board,
            final Activity activity,
            final Context context) {
        this.messenger = messenger;
        this.board = board;
        this.activity = activity;

        board.onUnexpectedDisconnect(new MetaWearBoard.UnexpectedDisconnectHandler() {
            @Override
            public void disconnected(int status) {
                stateHandler.success(
                        new HashMap<String, Object>() {
                            {
                                put("connected", false);
                                put("reason", "disconnected");
                            }
                        });
                clearHandlers();
            }
        });

        deviceChannel = new MethodChannel(messenger, getRootNamespace());
        deviceChannel.setMethodCallHandler(this);

        Log.d(MetawearDartPlugin.TAG, "MetawearBoardChannel: " + getRootNamespace());

        stateHandler = new StreamHandler(context);
        new EventChannel(messenger, getRootNamespace() + "/state").setStreamHandler(stateHandler);
    }

    public String getRootNamespace() {
        return MetawearDartPlugin.NAMESPACE + "/metawear/" + board.getMacAddress();
    }

    public MetaWearBoard getBoard() {
        return board;
    }

    private void clearHandlers() {
        messenger.setMessageHandler(getRootNamespace(), null);
        // registrar.messenger().setMessageHandler(getRootNamespace() + "/modules",
        // null);
    }

    @Override
    public void onMethodCall(final MethodCall methodCall, final MethodChannel.Result result) {
        final MetawearBoardChannel metawearChannel = this;
        Log.d(MetawearDartPlugin.TAG, "onMethodCall: " + methodCall.method + " " + methodCall.arguments);

        switch (methodCall.method) {
            case "connect": {
                Log.d(MetawearDartPlugin.TAG, "connect");
                board.connectAsync().continueWith(new Continuation<Void, Object>() {
                    @Override
                    public Object then(Task<Void> task) throws Exception {
                        if (task.isFaulted()) {
                            result.error("connect", task.getError().getMessage(), null);
                        } else {
                            stateHandler.success(
                                    new HashMap<String, Object>() {
                                        {
                                            put("connected", true);
                                        }
                                    });
                            result.success(true);
                        }
                        return null;
                    }
                });
            }
                break;
            case "disconnect": {
                Log.d(MetawearDartPlugin.TAG, "disconnect");
                board.disconnectAsync().continueWith(new Continuation<Void, Object>() {
                    @Override
                    public Object then(Task<Void> task) throws Exception {
                        result.success(true);
                        clearHandlers();
                        return null;
                    }
                });
            }
                break;
            case "model":
                result.success(board.getModel().toString());
                break;
            case "deviceInfo":
                board.readDeviceInformationAsync().continueWith(new Continuation<DeviceInformation, Object>() {
                    @Override
                    public Object then(Task<DeviceInformation> task) throws Exception {
                        final DeviceInformation deviceInformation = task.getResult();
                        result.success(new HashMap<String, Object>() {
                            {
                                put("manufacturer", deviceInformation.manufacturer);
                                put("modelNumber", deviceInformation.modelNumber);
                                put("serialNumber", deviceInformation.serialNumber);
                                put("firmwareRevision", deviceInformation.firmwareRevision);
                                put("hardwareRevision", deviceInformation.hardwareRevision);
                            }
                        });
                        return null;
                    }
                });
                break;
            case "battery":
                board.readBatteryLevelAsync().continueWith(new Continuation<Byte, Object>() {
                    @Override
                    public Object then(Task<Byte> task) throws Exception {
                        result.success(task.getResult());
                        return null;
                    }
                });
                break;
            case "startCorrectedAcceleration":
                SensorFusionBosch s = this.getSensor(methodCall);
                s.correctedAcceleration()
                        .addRouteAsync(source1 -> source1.stream((data, env) -> {
                            this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    float x = data.value(SensorFusionBosch.CorrectedAcceleration.class).x();
                                    float y = data.value(SensorFusionBosch.CorrectedAcceleration.class).y();
                                    float z = data.value(SensorFusionBosch.CorrectedAcceleration.class).z();
                                    long time = data.timestamp().getTimeInMillis();
                                    deviceChannel.invokeMethod("onCorrectedAcceleration",
                                            new HashMap<String, Object>() {
                                                {
                                                    put("x", x);
                                                    put("y", y);
                                                    put("z", z);
                                                    put("time", time);
                                                }

                                            ;
                                            });

                                }
                            });
                        }));
                s.correctedAcceleration().start();
                s.start();
                result.success(true);
                break;
            case "startCorrectedAngularVelocity":
                s = this.getSensor(methodCall);
                s.correctedAngularVelocity()
                        .addRouteAsync(source1 -> source1.stream((data, env) -> {

                            this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    float x = data.value(SensorFusionBosch.CorrectedAngularVelocity.class).x();
                                    float y = data.value(SensorFusionBosch.CorrectedAngularVelocity.class).y();
                                    float z = data.value(SensorFusionBosch.CorrectedAngularVelocity.class).z();
                                    long time = data.timestamp().getTimeInMillis();
                                    deviceChannel.invokeMethod("onCorrectedAngularVelocity",
                                            new HashMap<String, Object>() {
                                                {
                                                    put("x", x);
                                                    put("y", y);
                                                    put("z", z);
                                                    put("time", time);
                                                }

                                            ;
                                            });

                                }
                            });
                        }));
                s.correctedAngularVelocity().start();
                s.start();
                result.success(true);
                break;
            case "startCorrectedMagneticField":
                s = this.getSensor(methodCall);
                sensor.correctedMagneticField()
                        .addRouteAsync(source1 -> source1.stream((data, env) -> {

                            this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    float x = data.value(SensorFusionBosch.CorrectedMagneticField.class).x();
                                    float y = data.value(SensorFusionBosch.CorrectedMagneticField.class).y();
                                    float z = data.value(SensorFusionBosch.CorrectedMagneticField.class).z();
                                    long time = data.timestamp().getTimeInMillis();
                                    deviceChannel.invokeMethod("onCorrectedMagneticField",
                                            new HashMap<String, Object>() {
                                                {
                                                    put("x", x);
                                                    put("y", y);
                                                    put("z", z);
                                                    put("time", time);
                                                }

                                            ;
                                            });

                                }
                            });
                        }));
                sensor.correctedMagneticField().start();
                sensor.start();
                result.success(true);
                break;
            case "startQuaternion":
                s = this.getSensor(methodCall);
                s.quaternion()
                        .addRouteAsync(source1 -> source1.stream((data, env) -> {

                            this.activity.runOnUiThread(new Runnable() {
                                @Override
                                public void run() {
                                    float w = data.value(Quaternion.class).w();
                                    float x = data.value(Quaternion.class).x();
                                    float y = data.value(Quaternion.class).y();
                                    float z = data.value(Quaternion.class).z();
                                    long time = data.timestamp().getTimeInMillis();
                                    deviceChannel.invokeMethod("onQuaternion",
                                            new HashMap<String, Object>() {
                                                {
                                                    put("w", w);
                                                    put("x", x);
                                                    put("y", y);
                                                    put("z", z);
                                                    put("time", time);
                                                }

                                            ;
                                            });

                                }
                            });
                        }));
                s.quaternion().start();
                s.start();
                result.success(true);
                break;
            case "stop":
                s = this.getSensor(methodCall);
                s.correctedAcceleration().stop();
                s.correctedAngularVelocity().stop();
                s.correctedMagneticField().stop();
                s.quaternion().stop();
                s.stop();
                result.success(true);
                break;
        }
    }

    public interface MetawearDispose {
        void onDispose();
    }

}
